addons/calendar/models/calendar.py

```diff
@@ -0,0 +1,1047 @@
+    def _compute_display_start(self):
+        for meeting in self:
+            meeting.display_start = meeting.start_date if meeting.allday else meeting.start_datetime
+
+    @api.depends('start_date', 'allday', 'start_datetime')
+    def _compute_start(self):
+        for meeting in self:
+            meeting.start = meeting.start_date if meeting.allday else meeting.start_datetime
+
+    @api.depends('stop_date', 'allday', 'stop_datetime')
+    def _compute_stop(self):
+        for meeting in self:
+            meeting.stop = meeting.stop_date if meeting.allday else meeting.stop_datetime
+
+    @api.constrains('start_datetime', 'stop_datetime', 'start_date', 'stop_date')
+    def _check_closing_date(self):
+        for event in self:
+            if event.stop < event.start:
+                raise ValidationError(_('Error ! End date cannot be set before start date.'))
+
+    @api.onchange('allday')
+    def onchange_allday(self):
+        if not ((self.start_date and self.stop_date) or (self.start and self.stop)):  # At first intialize, we have not datetime
+            return
+
+        if self.allday:  # from datetime to date
+            start_datetime = self.start_datetime or self.start
+            if start_datetime:
+                start = fields.Datetime.from_string(start_datetime)
+                self.start_date = fields.Datetime.to_string(start)
+
+            stop_datetime = self.stop_datetime or self.stop
+            if stop_datetime:
+                stop = fields.Datetime.from_string(stop_datetime)
+                self.stop_date = fields.Datetime.to_string(stop)
+        else:  # from date to datetime
+            tz = pytz.timezone(self.env.user.tz) if self.env.user.tz else pytz.utc
+
+            if self.start_date:
+                start = fields.Datetime.from_string(self.start_date)
+                startdate = tz.localize(start).replace(hour=8).astimezone(pytz.utc)  # Add "+hh:mm" timezone, Set 8 AM in localtime, Convert to UTC
+                self.start_datetime = fields.Datetime.to_string(startdate)
+            elif self.start:
+                self.start_datetime = self.start
+
+            if self.stop_date:
+                stop = fields.Datetime.from_string(self.stop_date)
+                stop_date = tz.localize(stop).replace(hour=18).astimezone(pytz.utc)
+
+                self.stop_datetime = fields.Datetime.to_string(stop_date)
+            elif self.stop:
+                self.stop_datetime = self.stop
+
+    @api.onchange('duration', 'start_datetime')
+    def onchange_duration(self):
+        if not (self.start_datetime and self.duration):
+            return
+        start = fields.Datetime.from_string(self.start_datetime)
+        self.stop_date = fields.Datetime.to_string((start + timedelta(hours=self.duration)))
+        self.stop = fields.Datetime.to_string((start + timedelta(hours=self.duration)))
+        self.stop_datetime = fields.Datetime.to_string((start + timedelta(hours=self.duration)))
```

Comment:
start = fields.Datetime.from_string(self.start_datetime) + timedelta(hours=self.duration) and then..
fields.Datetime/Date.to_string(start)